lib/cartopy/mpl/patch.py

```diff
@@ -249,3 +265,183 @@ def not_zero_poly(geom):
+
+    # Figure out what type of object to return
+    if not polygons:
+        if not linestrings:
+            if not points:
+                # No geometries.  Return an empty point
+                return sgeom.Point()
+            elif len(points) > 1:
+                return sgeom.MultiPoint(points)
+            else:
+                return points[0]
+        elif not points:
+            if len(linestrings) > 1:
+                return sgeom.MultiLineString(linestrings)
+            else:
+                return linestrings[0]
+    else:
+        if not linestrings and not points:
+            if len(polygons) > 1:
+                return sgeom.MultiPolygon(polygons)
+            else:
+                return polygons[0]
+
+    # If we got to here, we have at least two types of geometry, so return
+    # a geometry collection.
+    return sgeom.GeometryCollection(polygons + linestrings + points)
+
+
+def shapely_to_path(shape):
+    """
+    Create a :class:`matplotlib.path.Path` object that describes a shape.
+
+    Parameters
+    ----------
+    shape
+        :class:`shapely.geometry.point.Point`,
+        :class:`shapely.geometry.linestring.LineString`,
+        :class:`shapely.geometry.polygon.LinearRing`,
+        :class:`shapely.geometry.polygon.Polygon`,
+        :class:`shapely.geometry.multipoint.MultiPoint`,
+        :class:`shapely.geometry.multipolygon.MultiPolygon`,
+        :class:`shapely.geometry.multilinestring.MultiLineString`,
+        :class:`shapely.geometry.collection.GeometryCollection`,
+        or any type with a _as_mpl_path() method.
+
+    Returns
+    -------
+    path
+        :class:`matplotlib.path.Path`
+
+    """
+    if shape.is_empty:
+        return Path(np.empty([0, 2]))
+    elif isinstance(shape, sgeom.LinearRing):
+        return Path(np.column_stack(shape.xy), closed=True)
+    elif isinstance(shape, (sgeom.LineString, sgeom.Point)):
+        return Path(np.column_stack(shape.xy))
+    elif isinstance(shape, sgeom.Polygon):
+        def poly_codes(poly):
+            codes = np.ones(len(poly.xy[0])) * Path.LINETO
+            codes[0] = Path.MOVETO
+            codes[-1] = Path.CLOSEPOLY
+            return codes
+        vertices = np.concatenate([np.array(shape.exterior.xy)] +
+                                  [np.array(ring.xy) for ring in
+                                   shape.interiors], 1).T
+        codes = np.concatenate([poly_codes(shape.exterior)] +
+                               [poly_codes(ring) for ring in shape.interiors])
+        return Path(vertices, codes)
+    elif isinstance(shape, (sgeom.MultiPolygon, sgeom.GeometryCollection,
+                            sgeom.MultiLineString, sgeom.MultiPoint)):
+        paths = []
+        for geom in shape.geoms:
+            path = shapely_to_path(geom)
+            if _MPL_38 or path.vertices.size > 0:
+                paths.append(path)
+        return Path.make_compound_path(*paths)
+    elif hasattr(shape, '_as_mpl_path'):
```

Comment:
What shapes/objects have `_as_mpl_path` that we haven't handled above? This seems a bit odd to have the function `shapely_to_path` but then accept other things besides shapely geometries.